Adding functionality for general queries plus more

Aryllen opened the pull request:
There's a lot here. Let me know, @kelshmo, if you want a meeting or have any questions. Also, I apologize if there's styling and whatnot out of place; feel free to be picky!

**Additions**

_Example_
- Add Rmarkdown example for updating publications for AD Knowledge Portal using grant serial numbers

_Text Cleaning_

- Add `hacky_cleaning` to clean up html formatting, non-ascii characters, and miscellaneous problems found
- Add `remove_unacceptable_characters` to clean up anything that isn't allowed (includes `hacky_cleaning`, but adds additional restrictions due to entity name requirements)
- Add `remove_html_formatting` to remove html formatting (used in `hacky_cleaning`)
- Add `convert_to_ascii` to convert non-ascii to ascii (used in `hacky_cleaning`)

_Queries_

- Add function, `query_list_general`, to gather Pubmed query info using a list of strings, where each string is a query
- Add `query_general` to gather just the Pubmed IDs for a single query (used in `query_list_general`)

**Changes**
- Add `stringr` package name when using `str_trunc` for author truncation.
- Truncate the author + journal section of the entity name to ensure that the entity name will be under 256 characters
- Add `sleep_time` to existing query functions to get around Pubmed rate limiting

kelshmo commented on R/functions.R:
```diff
@@ -49,26 +61,72 @@ collapse_pm_query <- function(df) {
+#' - abbreviates the author names by first initial, last name
+#' - creates one row per PubmedId
+#' - query without associated PubmedIds will be left out of returned data
+#' - query added as an additional column
```
```suggestion
#' - query added as an additional column
```
What does this mean?

Aryllen replied:
There's an extra column in the dataframe that has the name 'query' and the string that was used as the Pubmed query.

For example, if I query using `c('foo', 'bar')`, I'll get back something like this:
| pmid  | journal | more columns | query |
| ------------- | ------------- | --- | --- |
| 12345678  | neat journal  | ... | foo |
| 88382848 | lame journal | ... | foo |
| 21324332  | cool journa  | ... | bar |

This helps with knowing what query returned a journal. In the AD case, they are using portions of the grant number as a query. I can map the full grant number to this table using the query column.

kelshmo commented on R/functions.R:
```diff
@@ -49,26 +61,72 @@ collapse_pm_query <- function(df) {
+}
+#' Query for Pubmed IDs only
+#'
+#' @param query Character string with query
```
```suggestion
#' @param query Character string with query
```
Could you provide an example query?

Aryllen replied:
Sure. I'll add that.

kelshmo commented on vignettes/query-pubmed-grants.Rmd:
```diff
@@ -0,0 +1,112 @@
+
+## Query Pubmed and store data as file annotations
+
+The data needed for these steps are *Grant Number*, *grantSerialNumber* and *Program*. Theses functions take a list of grant serial numbers and queries the site to pull down title, abstract, authors, journal name, year and DOI. Theses annotations are visible in the [AD Knowledge Portal - Publications View](https://www.synapse.org/#!Synapse:syn20448807/tables/). See the [Explore Publications modul ...
```
```suggestion
The data needed for these steps are *Grant Number*, *grantSerialNumber* and *Program*. Theses functions take a list of grant serial numbers and queries Pubmed to download title, abstract, authors, journal name, year and DOI. Theses annotations are visible in the [AD Knowledge Portal - Publications View](https://www.synapse.org/#!Synapse:syn20448807/tables/). See the [Explore Publications module](https://adknowledgeportal.synapse.org/Explore/Publications) for a visual of how this data is surfaced on the portal.
```

kelshmo commented on vignettes/query-pubmed-grants.Rmd:
````diff
@@ -0,0 +1,112 @@
+store_as_annotations(parent = parent, dat_list)
+```
+
+Query the publications table to force an update.
````
nice!

kelshmo reviewed (changes requested):
Overall, LG!

I have small suggestions to clarify verbiage here and there. 

The only truly requested change is to add a check in `make_entity_name()` to make sure the final product does not exceed the character limit.

Aryllen commented on R/functions.R:
```diff
@@ -5,22 +5,34 @@
+    glue::glue("{first$lastname} {first$journal}"),
+    width = 200
+  )
+  name <- glue::glue("{short_name} {first$year} (Pubmed ID {first$pmid})")
```
That could be added. The main reason why I didn't is that years are always 4 characters (at least for the relevant future) and the Pubmed ID is generally 8 characters. These two things added with the extra formatting take up 26 characters. Since I cut down the first bit to 200 characters, that leaves about 36 characters of buffer, in case Pubmed increases the length of their ids.

If you would like an extra check here, what would you like it to look like? Another truncation to 256 or throwing an error (i.e. stopping that one from finishing)?

kelshmo replied:
You make a great point! No need to add that check.

Aryllen commented:
The DESCRIPTION was missing a few imports. I added those in the latest updates.

kelshmo commented:
@Aryllen marked this as draft since I think you are adding more to this PR? If I'm mistaken, and you plan to open another PR, just mark as ready :)

Aryllen commented:
Added script for AD portal publication update.

Aryllen commented:
This won't let me change it back to a draft. There's some hard-coded values (task view, table ids, etc) in here since this script is super specific to AD publications. That said, would you prefer that I add those as command line arguments to the script?

kelshmo commented:
That would be SO helpful so I can set this up to automate!

Aryllen commented:
This is ready for review, again, @kelshmo! The script now takes more arguments. When I add the bio/medRxiv publication info, I might want to revisit how some of this is set up (make it more flexible), but for now, it does the job.

I tested this on the server to make sure it worked.

Aryllen commented:
@kelshmo, can you please take a look at this? I've updated the publications-query-ad.R to work better with a GH Action. I've added a GH action to use it, as well. I'm not completely sure it's correct, though. In the interest of time, I've tried mashing up a couple repo's actions (synapseAnnotations and dccvalidator) to try to hit all the needed elements. It should run on Sunday at 10pm PST, if I have UTC to PST correct. If it failed, you'd get a notification that the action failed. It _shouldn't_ hurt anything if it fails since it would likely just fail the moment it tried running the script and/or logging into Synapse. Would just need to troubleshoot the Action setup. The action would need a Synapse PAT (called SYNAPSE_PAT) as a secret. I don't have the ability to add that. I'd recommend generating a PAT from the shared sysbio-dcc Synapse service account. It will need all the permissions in order to update the publications.

There is also an Rmd I added that has info on how to generate a table of publications from an expanded PubMed query with an indication of whether a Sage-supported PI is in the author list.

kelshmo reviewed (approved):
PAT added. Your changes look good. I will keep on eye on Sunday :)
